File: platform_testing/libraries/flag-helpers/junit/src_host/android/platform/test/flag/junit/host/DeviceFlags.java
/*
 * Copyright (C) 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package android.platform.test.flag.junit.host;

import android.aconfig.Aconfig.flag_permission;
import android.aconfig.Aconfig.flag_state;
import android.aconfig.Aconfig.parsed_flags;
import android.platform.test.flag.util.Flag;
import android.platform.test.flag.util.FlagReadException;

import com.android.tradefed.device.DeviceNotAvailableException;
import com.android.tradefed.device.ITestDevice;
import com.android.tradefed.log.LogUtil;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.Nullable;

/**
 * Dumps flags from a give device.
 *
 * <p>There are two sources to dump the flag values: device_config and aconfig dumped output. The
 * device_config contains the ground truth for readwrite flags (including legacy flags), while the
 * aconfig dumped output contains the ground truth for all readonly flags.
 *
 * <p>The {packageName}.{flagName} (we call it full flag name) is the unique identity of each
 * aconfig flag, which is also the constant value for each flag name in the aconfig auto generated
 * library. However, each aconfig flag definitions also contains namespace which is in aconfig
 * dumped output. The {@code DeviceFlags} saves all flag values into a map with the key always
 * starts with "{namespace}/" no matter it is a legacy flag or aconfig flag, and gets flag values
 * from this map.
 */
public class DeviceFlags {
    private static final String DUMP_DEVICE_CONFIG_CMD = "device_config list";

    /**
     * Partitions that contain the aconfig_flags.pb files. Should be consistent with the partitions
     * defined in core/packaging/flags.mk.
     */
    private static final List<String> FLAG_PARTITIONS =
            List.of("product", "system", "system_ext", "vendor");

    /**
     * The key is the flag name with namespace ({namespace}/{flagName} for legacy flags,
     * {namespace}/{packageName}.{flagName} for aconfig flags.
     */
    private final Map<String, String> mAllFlagValues = new HashMap<>();

    /**
     * The key is the aconfig full flag name ({packageName}.{flagName}), the value is the flag name
     * with namespace ({namespace}/{packageName}.{flagName}).
     */
    private final Map<String, String> mFlagNameWithNamespaces = new HashMap<>();

    private DeviceFlags() {}

    public static DeviceFlags createDeviceFlags(ITestDevice testDevice) throws FlagReadException {
        DeviceFlags deviceFlags = new DeviceFlags();
        deviceFlags.init(testDevice);
        return deviceFlags;
    }

    @Nullable
    public String getFlagValue(String flag) {
        String flagWithNamespace =
                flag.contains(Flag.NAMESPACE_FLAG_SEPARATOR)
                        ? flag
                        : mFlagNameWithNamespaces.get(flag);

        return mAllFlagValues.get(flagWithNamespace);
    }

    public void init(ITestDevice testDevice) throws FlagReadException {
        mAllFlagValues.clear();
        mFlagNameWithNamespaces.clear();
        LogUtil.CLog.i("Dumping all flag values from the device ...");
        getDeviceConfigFlags(testDevice);
        getAconfigFlags(testDevice);
        LogUtil.CLog.i("Dumped all flag values from the device.");
    }

    private void getDeviceConfigFlags(ITestDevice testDevice) throws FlagReadException {
        try {
            String deviceConfigList = testDevice.executeShellCommand(DUMP_DEVICE_CONFIG_CMD);
            for (String deviceConfig : deviceConfigList.split("\n")) {
                if (deviceConfig.contains("=")) {
                    String[] flagNameValue = deviceConfig.split("=", 2);
                    mAllFlagValues.put(flagNameValue[0], flagNameValue[1]);
                } else {
                    LogUtil.CLog.w("Invalid device config %s", deviceConfig);
                }
            }
        } catch (DeviceNotAvailableException e) {
            throw new FlagReadException("ALL_FLAGS", e);
        }
    }

    private void getAconfigFlags(ITestDevice testDevice) throws FlagReadException {
        getAconfigParsedFlags(testDevice)
                .getParsedFlagList()
                .forEach(
                        parsedFlag -> {
                            String namespace = parsedFlag.getNamespace();
                            String fullFlagName =
                                    String.format(
                                            Flag.ACONFIG_FULL_FLAG_FORMAT,
                                            parsedFlag.getPackage(),
                                            parsedFlag.getName());
                            String flagWithNamespace =
                                    String.format(
                                            Flag.FLAG_WITH_NAMESPACE_FORMAT,
                                            namespace,
                                            fullFlagName);
                            mFlagNameWithNamespaces.put(fullFlagName, flagWithNamespace);
                            String defaultValue =
                                    parsedFlag.getState().equals(flag_state.ENABLED)
                                            ? "true"
                                            : "false";
                            if (parsedFlag.getPermission().equals(flag_permission.READ_ONLY)
                                    || !mAllFlagValues.containsKey(flagWithNamespace)) {
                                mAllFlagValues.put(flagWithNamespace, defaultValue);
                            }
                        });
    }

    private parsed_flags getAconfigParsedFlags(ITestDevice testDevice) throws FlagReadException {
        parsed_flags.Builder builder = parsed_flags.newBuilder();

        for (String flagPartition : FLAG_PARTITIONS) {
            try {
                String aconfigFlagsPbFilePath =
                        String.format("/%s/etc/aconfig_flags.pb", flagPartition);
                if (!testDevice.doesFileExist(aconfigFlagsPbFilePath)) {
                    LogUtil.CLog.i("Aconfig flags file %s does not exist", aconfigFlagsPbFilePath);
                    continue;
                }
                File aconfigFlagsPbFile = testDevice.pullFile(aconfigFlagsPbFilePath);
                if (aconfigFlagsPbFile.length() <= 1) {
                    LogUtil.CLog.i("Aconfig flags file %s is empty", aconfigFlagsPbFilePath);
                    continue;
                }
                InputStream aconfigFlagsPbInputStream = new FileInputStream(aconfigFlagsPbFile);
                parsed_flags parsedFlags = parsed_flags.parseFrom(aconfigFlagsPbInputStream);
                builder.addAllParsedFlag(parsedFlags.getParsedFlagList());
            } catch (DeviceNotAvailableException | IOException e) {
                throw new FlagReadException("ALL_FLAGS", e);
            }
        }
        return builder.build();
    }
}

File: tools/test/mobly_extensions/tools/device_flags.py
#!/usr/bin/env python3

# Copyright (C) 2023 The Android Open Source Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Utility for accessing aconfig and device_config flag values on a device."""

import os
import tempfile
from typing import Any, Dict, Optional

from mobly.controllers import android_device
from protos import aconfig_pb2

_ACONFIG_PARTITIONS = ('product', 'system', 'system_ext', 'vendor')
_ACONFIG_PB_FILE = 'aconfig_flags.pb'

_DEVICE_CONFIG_GET_CMD = 'device_config get'

_READ_ONLY = aconfig_pb2.flag_permission.READ_ONLY
_ENABLED = aconfig_pb2.flag_state.ENABLED


class DeviceFlags:
    """Provides access to aconfig and device_config flag values of a device."""

    def __init__(self, ad: android_device.AndroidDevice):
        self._ad = ad
        self._aconfig_flags = None

    def get_value(self, namespace: str, name: str) -> Optional[str]:
        """Gets the value of the requested flag.

        Flags must be specified by both its namespace and name.

        The method will first look for the flag from the device's
        aconfig_flags.pb files, and, if not found or the flag is READ_WRITE,
        then retrieve the value from 'adb device_config get'.

        All values are returned as strings, e.g. 'true', '3'.

        Args:
            namespace: The namespace of the flag.
            name: The full name of the flag.
                For aconfig flags, it is equivalent to '{package}.{name}' from
                    the aconfig proto.
                For device_config flags, it is equivalent to '{KEY}' from the
                    "device_config shell" command.

        Returns:
            The flag value as a string.
        """
        # Check aconfig
        aconfig_val = None
        aconfig_flag = self._get_aconfig_flags().get(
            '%s/%s' % (namespace, name))
        if aconfig_flag is not None:
            aconfig_val = 'true' if aconfig_flag.state == _ENABLED else 'false'
            if aconfig_flag.permission == _READ_ONLY:
                return aconfig_val

        # If missing or READ_WRITE, also check device_config
        device_config_val = self._ad.adb.shell(
            '%s %s %s' % (_DEVICE_CONFIG_GET_CMD, namespace, name)
        ).decode('utf8').strip()
        return device_config_val if device_config_val != 'null' else aconfig_val

    def get_bool(self, namespace: str, name: str) -> bool:
        """Gets the value of the requested flag as a boolean.

        See get_value() for details.

        Args:
            namespace: The namespace of the flag.
            name: The key of the flag.

        Returns:
            The flag value as a boolean.

        Raises:
            ValueError if the flag value cannot be expressed as a boolean.
        """
        val = self.get_value(namespace, name)
        if val.lower() == 'true':
            return True
        if val.lower() == 'false':
            return False
        raise ValueError('Flag %s/%s is not a boolean (value: %s).'
                         % (namespace, name, val))

    def _get_aconfig_flags(self) -> Dict[str, Any]:
        """Gets the aconfig flags as a dict. Loads from proto if necessary."""
        if self._aconfig_flags is None:
            self._load_aconfig_flags()
        return self._aconfig_flags

    def _load_aconfig_flags(self) -> None:
        """Pull aconfig proto files from device, then load the flag info."""
        self._aconfig_flags = {}
        with tempfile.TemporaryDirectory() as tmp_dir:
            for partition in _ACONFIG_PARTITIONS:
                device_path = os.path.join(
                    '/', partition, 'etc', _ACONFIG_PB_FILE)
                host_path = os.path.join(
                    tmp_dir, '%s_%s' % (partition, _ACONFIG_PB_FILE))
                self._ad.adb.pull([device_path, host_path])
                with open(host_path, 'rb') as f:
                    parsed_flags = aconfig_pb2.parsed_flags.FromString(f.read())
                for flag in parsed_flags.parsed_flag:
                    full_name = '%s/%s.%s' % (
                        flag.namespace, flag.package, flag.name)
                    self._aconfig_flags[full_name] = flag
